# Lab | Web Scraping

Welcome to the "Books to Scrape" Web Scraping Adventure Lab!

**Objective**

In this lab, we will embark on a mission to unearth valuable insights from the data available on Books to Scrape, an online platform showcasing a wide variety of books. As data analyst, you have been tasked with scraping a specific subset of book data from Books to Scrape to assist publishing companies in understanding the landscape of highly-rated books across different genres. Your insights will help shape future book marketing strategies and publishing decisions.

**Background**

In a world where data has become the new currency, businesses are leveraging big data to make informed decisions that drive success and profitability. The publishing industry, much like others, utilizes data analytics to understand market trends, reader preferences, and the performance of books based on factors such as genre, author, and ratings. Books to Scrape serves as a rich source of such data, offering detailed information about a diverse range of books, making it an ideal platform for extracting insights to aid in informed decision-making within the literary world.

**Task**

Your task is to create a Python script using BeautifulSoup and pandas to scrape Books to Scrape book data, focusing on book ratings and genres. The script should be able to filter books with ratings above a certain threshold and in specific genres. Additionally, the script should structure the scraped data in a tabular format using pandas for further analysis.

**Expected Outcome**

A function named `scrape_books` that takes two parameters: `min_rating` and `max_price`. The function should scrape book data from the "Books to Scrape" website and return a `pandas` DataFrame with the following columns:

**Expected Outcome**

- A function named `scrape_books` that takes two parameters: `min_rating` and `max_price`.
- The function should return a DataFrame with the following columns:
  - **UPC**: The Universal Product Code (UPC) of the book.
  - **Title**: The title of the book.
  - **Price (£)**: The price of the book in pounds.
  - **Rating**: The rating of the book (1-5 stars).
  - **Genre**: The genre of the book.
  - **Availability**: Whether the book is in stock or not.
  - **Description**: A brief description or product description of the book (if available).
  
You will execute this script to scrape data for books with a minimum rating of `4.0 and above` and a maximum price of `£20`. 

Remember to experiment with different ratings and prices to ensure your code is versatile and can handle various searches effectively!

**Resources**

- [Beautiful Soup Documentation](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)
- [Pandas Documentation](https://pandas.pydata.org/pandas-docs/stable/index.html)
- [Books to Scrape](https://books.toscrape.com/)


**Hint**

Your first mission is to familiarize yourself with the **Books to Scrape** website. Navigate to [Books to Scrape](http://books.toscrape.com/) and explore the available books to understand their layout and structure. 

Next, think about how you can set parameters for your data extraction:

- **Minimum Rating**: Focus on books with a rating of 4.0 and above.
- **Maximum Price**: Filter for books priced up to £20.

After reviewing the site, you can construct a plan for scraping relevant data. Pay attention to the details displayed for each book, including the title, price, rating, and availability. This will help you identify the correct HTML elements to target with your scraping script.

Make sure to build your scraping URL and logic based on the patterns you observe in the HTML structure of the book listings!


---

**Best of luck! Immerse yourself in the world of books, and may the data be with you!**

**Important Note**:

In the fast-changing online world, websites often update and change their structures. When you try this lab, the **Books to Scrape** website might differ from what you expect.

If you encounter issues due to these changes, like new rules or obstacles preventing data extraction, don’t worry! Get creative.

You can choose another website that interests you and is suitable for scraping data. Options like Wikipedia, The New York Times, or even library databases are great alternatives. The main goal remains the same: extract useful data and enhance your web scraping skills while exploring a source of information you enjoy. This is your opportunity to practice and adapt to different web environments!

In [2]:
%pip install requests beautifulsoup4 pandas

Defaulting to user installation because normal site-packages is not writeable
  Using cached requests-2.32.5-py3-none-any.whl (64 kB)
     |████████████████████████████████| 109 kB 3.8 MB/s eta 0:00:01
     |████████████████████████████████| 68 kB 2.4 MB/s eta 0:00:01
  Using cached urllib3-2.6.3-py3-none-any.whl (131 kB)
     |████████████████████████████████| 136 kB 2.3 MB/s eta 0:00:01
     |████████████████████████████████| 368 kB 3.6 MB/s eta 0:00:01
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


## 1. Import libraries and connect to the website

We use:

- `requests` to download the website's HTML.
- `BeautifulSoup` to read and search the HTML structure.
- `pandas` to store the scraped information in a DataFrame.

The status code `200` confirms that the website responded successfully.

In [3]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

url = "https://books.toscrape.com/"

response = requests.get(url)

print(response.status_code)

/Users/sozdemirdeni/Library/Python/3.9/lib/python/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(


200


## 2. Read the HTML and find the books

The website stores each book inside an HTML element called
`article.product_pod`.

For the first book, we extract:

- The title from the `title` attribute.
- The price from the element with class `price_color`.
- The availability from the element with class `availability`.

There are 20 books displayed on each webpage.

In [5]:
soup = BeautifulSoup(response.text, "html.parser")

book_cards = soup.select("article.product_pod")

print(f"Number of books on this page: {len(book_cards)}")

first_book = book_cards[0]

title = first_book.h3.a["title"]
price = first_book.select_one(".price_color").get_text(strip=True)
availability = first_book.select_one(".availability").get_text(" ", strip=True)

print("Title:", title)
print("Price:", price)
print("Availability:", availability)

Number of books on this page: 20
Title: A Light in the Attic
Price: £51.77
Availability: In stock


## 3. Extract and clean rating and price

The price is shown as text, for example `£51.77`. We remove the pound
symbol and convert the remaining value to a decimal number.

The rating is stored as a CSS class, such as `Four` or `Five`, rather
than as a numerical value. We use a dictionary to convert the rating
word into a number.

In [7]:
response.encoding = "utf-8"

soup = BeautifulSoup(response.text, "html.parser")
book_cards = soup.select("article.product_pod")

rating_mapping = {
    "One": 1,
    "Two": 2,
    "Three": 3,
    "Four": 4,
    "Five": 5
}

first_book = book_cards[0]

price_text = first_book.select_one(".price_color").get_text(strip=True)
price = float(price_text.replace("£", ""))

rating_text = first_book.select_one("p.star-rating")["class"][1]
rating = rating_mapping[rating_text]

print("Price:", price)
print("Rating:", rating)

Price: 51.77
Rating: 3


## 4. Extract all books from the current page

We loop through all book cards on the page and extract the title, price,
rating, and availability.

Each book is stored as a dictionary. The dictionaries are then combined
into a pandas DataFrame, which gives us a tabular structure for further
analysis.

In [8]:
books_data = []

for book in book_cards:
    title = book.h3.a["title"]

    price = float(
        book.select_one(".price_color")
        .get_text(strip=True)
        .replace("£", "")
    )

    rating_text = book.select_one("p.star-rating")["class"][1]
    rating = rating_mapping[rating_text]

    availability = book.select_one(".availability").get_text(" ", strip=True)

    books_data.append({
        "Title": title,
        "Price (£)": price,
        "Rating": rating,
        "Availability": availability
    })

books_page_df = pd.DataFrame(books_data)

books_page_df.head()

,Title,Price (£),Rating,Availability
0,A Light in the Attic,51.77,3,In stock
1,Tipping the Velvet,53.74,1,In stock
2,Soumission,50.10,1,In stock
3,Sharp Objects,47.82,4,In stock
4,Sapiens: A Brief History of Humankind,54.23,5,In stock


## 5. Open a book's detail page

The book overview page does not contain all the required information.
Therefore, we use the link stored in each book card to visit the
individual product page.

From the detail page, we extract:

- The UPC from the product information table.
- The genre from the breadcrumb navigation.
- The description from the Product Description section.

`urljoin()` combines the main website URL with the relative book link.

In [10]:
from urllib.parse import urljoin

first_book = book_cards[0]

relative_url = first_book.h3.a["href"]
detail_url = urljoin(url, relative_url)

detail_response = requests.get(detail_url)
detail_response.encoding = "utf-8"

detail_soup = BeautifulSoup(detail_response.text, "html.parser")

# Extract the UPC and other product information
product_info = {}

for row in detail_soup.select("table.table-striped tr"):
    key = row.th.get_text(strip=True)
    value = row.td.get_text(strip=True)
    product_info[key] = value

# Extract the genre from the breadcrumb
breadcrumb_items = detail_soup.select("ul.breadcrumb li")
genre = breadcrumb_items[-2].get_text(strip=True)

# Extract the description
description_heading = detail_soup.find(
    "h2",
    string=lambda text: text and "Product Description" in text
)

if description_heading:
    description_paragraph = description_heading.find_next("p")
    description = description_paragraph.get_text(strip=True)
else:
    description = None

print("UPC:", product_info["UPC"])
print("Genre:", genre)
print("Description:", description)

UPC: a897fe39b1053632
Genre: Poetry
Description: It's hard to imagine a world without A Light in the Attic. This now-classic collection of poetry and drawings from Shel Silverstein celebrates its 20th anniversary with this special edition. Silverstein's humorous and creative verse can amuse the dowdiest of readers. Lemon-faced adults and fidgety kids sit still and read these rhythmic words and laugh and smile and love th It's hard to imagine a world without A Light in the Attic. This now-classic collection of poetry and drawings from Shel Silverstein celebrates its 20th anniversary with this special edition. Silverstein's humorous and creative verse can amuse the dowdiest of readers. Lemon-faced adults and fidgety kids sit still and read these rhythmic words and laugh and smile and love that Silverstein. Need proof of his genius? RockabyeRockabye baby, in the treetopDon't you know a treetopIs no safe place to rock?And who put you up there,And your cradle, too?Baby, I think someone down

## 6. Create a reusable function for one book

We place the extraction logic inside `scrape_one_book()`.

The function:

1. Receives one book card.
2. Extracts information from the overview page.
3. Opens the book's detail page.
4. Extracts the UPC, genre, and description.
5. Returns all information as a dictionary.

Using a function avoids repeating the same code for every book.

In [22]:
def scrape_one_book(book_card, base_url):
    # Basic information from the overview page
    title = book_card.h3.a["title"]

    price = float(
        book_card.select_one(".price_color")
        .get_text(strip=True)
        .replace("£", "")
    )

    rating_text = book_card.select_one(
        "p.star-rating"
    )["class"][1]

    rating = rating_mapping[rating_text]

    availability = book_card.select_one(
        ".availability"
    ).get_text(" ", strip=True)

    # Open the individual book page
    relative_url = book_card.h3.a["href"]

    if relative_url.startswith("../"):
        detail_url = urljoin(
            "https://books.toscrape.com/catalogue/",
            relative_url[3:]
        )
    else:
        detail_url = urljoin(base_url, relative_url)

    detail_response = requests.get(detail_url)
    detail_response.encoding = "utf-8"

    detail_soup = BeautifulSoup(
        detail_response.text,
        "html.parser"
    )

    # Extract product information
    product_info = {}

    for row in detail_soup.select(
        "table.table-striped tr"
    ):
        key = row.th.get_text(strip=True)
        value = row.td.get_text(strip=True)
        product_info[key] = value

    # Extract genre
    breadcrumb_items = detail_soup.select(
        "ul.breadcrumb li"
    )

    if len(breadcrumb_items) >= 2:
        genre = breadcrumb_items[-2].get_text(
            strip=True
        )
    else:
        genre = None

    # Extract description
    description_heading = detail_soup.find(
        "h2",
        string=lambda text: (
            text and "Product Description" in text
        )
    )

    if description_heading:
        description_paragraph = (
            description_heading.find_next("p")
        )

        if description_paragraph:
            description = description_paragraph.get_text(
                strip=True
            )
        else:
            description = None
    else:
        description = None

    return {
        "UPC": product_info.get("UPC"),
        "Title": title,
        "Price (£)": price,
        "Rating": rating,
        "Genre": genre,
        "Availability": availability,
        "Description": description
    }

In [13]:
first_book_data = scrape_one_book(book_cards[0], url)

first_book_data

{'UPC': 'a897fe39b1053632',
 'Title': 'A Light in the Attic',
 'Price (£)': 51.77,
 'Rating': 3,
 'Genre': 'Poetry',
 'Availability': 'In stock',
 'Description': "It's hard to imagine a world without A Light in the Attic. This now-classic collection of poetry and drawings from Shel Silverstein celebrates its 20th anniversary with this special edition. Silverstein's humorous and creative verse can amuse the dowdiest of readers. Lemon-faced adults and fidgety kids sit still and read these rhythmic words and laugh and smile and love th It's hard to imagine a world without A Light in the Attic. This now-classic collection of poetry and drawings from Shel Silverstein celebrates its 20th anniversary with this special edition. Silverstein's humorous and creative verse can amuse the dowdiest of readers. Lemon-faced adults and fidgety kids sit still and read these rhythmic words and laugh and smile and love that Silverstein. Need proof of his genius? RockabyeRockabye baby, in the treetopDon't y

## 7. Create a DataFrame for the current page

We apply the `scrape_one_book()` function to each book card on the page.

Each result is added to a list and then converted into a pandas
DataFrame. This DataFrame contains all seven required columns for the
20 books on the current page.

In [15]:
page_books_data = []

for book_card in book_cards:
    book_data = scrape_one_book(book_card, url)
    page_books_data.append(book_data)

page_df = pd.DataFrame(page_books_data)

print(page_df.shape)
page_df.head()

(20, 7)


,UPC,Title,Price (£),Rating,Genre,Availability,Description
0,a897fe39b1053632,A Light in the Attic,51.77,3,Poetry,In stock,It's hard to imagine a world without A Light i...
1,90fa61229261140a,Tipping the Velvet,53.74,1,Historical Fiction,In stock,"""Erotic and absorbing...Written with starling ..."
2,6957f44c3847a760,Soumission,50.10,1,Fiction,In stock,"Dans une France assez proche de la nôtre, un h..."
3,e00eb4fd7b871a48,Sharp Objects,47.82,4,Mystery,In stock,"WICKED above her hipbone, GIRL across her hear..."
4,4165285e1663650f,Sapiens: A Brief History of Humankind,54.23,5,History,In stock,From a renowned historian comes a groundbreaki...


## 8. Scrape all website pages

Books to Scrape displays 20 books per page and uses a `next` link for
pagination.

The `while` loop:

1. Downloads the current page.
2. Extracts all book cards.
3. Adds them to `all_book_cards`.
4. Finds the next-page link.
5. Stops when there is no next-page link.

This gives us access to all books on the website.

In [17]:
all_book_cards = []

page_url = url
page_number = 1

while page_url:
    page_response = requests.get(page_url)
    page_response.encoding = "utf-8"

    page_soup = BeautifulSoup(
        page_response.text,
        "html.parser"
    )

    current_page_cards = page_soup.select(
        "article.product_pod"
    )

    all_book_cards.extend(current_page_cards)

    next_link = page_soup.select_one("li.next a")

    if next_link:
        page_url = urljoin(
            page_url,
            next_link["href"]
        )
        page_number += 1
    else:
        page_url = None

print("Pages scraped:", page_number)
print("Books found:", len(all_book_cards))

Pages scraped: 50
Books found: 1000


## 9. Scrape the details for all books

We loop through all book cards and use the reusable
`scrape_one_book()` function.

The results are stored in a list of dictionaries and converted into a
DataFrame containing the seven required columns.

In [23]:
all_books_data = []

for number, book_card in enumerate(all_book_cards, start=1):
    book_data = scrape_one_book(book_card, url)
    all_books_data.append(book_data)

    if number % 100 == 0:
        print(f"{number} books scraped")

all_books_df = pd.DataFrame(all_books_data)

print("Final shape:", all_books_df.shape)

100 books scraped
200 books scraped
300 books scraped
400 books scraped
500 books scraped
600 books scraped
700 books scraped
800 books scraped
900 books scraped
1000 books scraped
Final shape: (1000, 7)


## 10. Filter the books

We select books that meet both requirements:

- Rating is 4 or higher.
- Price is £20 or lower.

The `&` operator means that both conditions must be true.
The `.copy()` method creates a separate filtered DataFrame.

In [24]:
filtered_books_df = all_books_df[
    (all_books_df["Rating"] >= 4) &
    (all_books_df["Price (£)"] <= 20)
].copy()

print("Number of matching books:", len(filtered_books_df))

filtered_books_df.head()

Number of matching books: 75


,UPC,Title,Price (£),Rating,Genre,Availability,Description
12,ce6396b0f23f6ecc,Set Me Free,17.46,5,Young Adult,In stock,Aaron Ledbetter’s future had been planned out ...
30,None,The Four Agreements: A Practical Guide to Pers...,17.66,5,None,In stock,None
34,None,Sophie's World,15.94,5,None,In stock,None
47,None,Untitled Collection: Sabbath Poems 2014,14.27,4,None,In stock,None
53,None,This One Summer,19.49,4,None,In stock,None


## 11. Create the required `scrape_books()` function

The `scrape_books()` function:

1. Scrapes all pages of the website.
2. Extracts the required information for every book.
3. Creates a pandas DataFrame.
4. Filters the DataFrame using `min_rating` and `max_price`.
5. Returns the filtered DataFrame.

The function is reusable because the filtering values are provided as
parameters.

In [25]:
def scrape_books(min_rating, max_price):
    all_book_cards = []
    page_url = "https://books.toscrape.com/"

    # Scrape all overview pages
    while page_url:
        page_response = requests.get(page_url)
        page_response.encoding = "utf-8"

        page_soup = BeautifulSoup(
            page_response.text,
            "html.parser"
        )

        current_page_cards = page_soup.select(
            "article.product_pod"
        )

        all_book_cards.extend(current_page_cards)

        next_link = page_soup.select_one("li.next a")

        if next_link:
            page_url = urljoin(
                page_url,
                next_link["href"]
            )
        else:
            page_url = None

    # Scrape the details for every book
    all_books_data = []

    for number, book_card in enumerate(
        all_book_cards,
        start=1
    ):
        book_data = scrape_one_book(
            book_card,
            "https://books.toscrape.com/"
        )

        all_books_data.append(book_data)

        if number % 100 == 0:
            print(f"{number} books scraped")

    books_df = pd.DataFrame(all_books_data)

    # Apply the requested filters
    filtered_df = books_df[
        (books_df["Rating"] >= min_rating) &
        (books_df["Price (£)"] <= max_price)
    ].copy()

    return filtered_df

In [26]:
filtered_books_df = all_books_df[
    (all_books_df["Rating"] >= 4) &
    (all_books_df["Price (£)"] <= 20)
].copy()

print("Matching books:", len(filtered_books_df))

filtered_books_df

Matching books: 75


,UPC,Title,Price (£),Rating,Genre,Availability,Description
12,ce6396b0f23f6ecc,Set Me Free,17.46,5,Young Adult,In stock,Aaron Ledbetter’s future had been planned out ...
30,None,The Four Agreements: A Practical Guide to Pers...,17.66,5,None,In stock,None
34,None,Sophie's World,15.94,5,None,In stock,None
47,None,Untitled Collection: Sabbath Poems 2014,14.27,4,None,In stock,None
53,None,This One Summer,19.49,4,None,In stock,None
...,...,...,...,...,...,...,...
913,None,The Zombie Room,19.69,5,None,In stock,None
917,None,The Silent Wife,12.34,5,None,In stock,None
934,None,The Girl You Lost,12.29,5,None,In stock,None
937,None,The Edge of Reason (Bridget Jones #2),19.18,4,None,In stock,None


In [27]:
final_result = scrape_books(
    min_rating=4.0,
    max_price=20
)

print("Final result shape:", final_result.shape)

final_result.head()

100 books scraped
200 books scraped
300 books scraped
400 books scraped
500 books scraped
600 books scraped
700 books scraped
800 books scraped
900 books scraped
1000 books scraped
Final result shape: (75, 7)


,UPC,Title,Price (£),Rating,Genre,Availability,Description
12,ce6396b0f23f6ecc,Set Me Free,17.46,5,Young Adult,In stock,Aaron Ledbetter’s future had been planned out ...
30,None,The Four Agreements: A Practical Guide to Pers...,17.66,5,None,In stock,None
34,None,Sophie's World,15.94,5,None,In stock,None
47,None,Untitled Collection: Sabbath Poems 2014,14.27,4,None,In stock,None
53,None,This One Summer,19.49,4,None,In stock,None
